concurrent_map.go

```diff
@@ -5,75 +5,114 @@ import (
 	"sync"
 )
 
-var SHARD_COUNT = 32
-
-// A "thread" safe map of type string:Anything.
-// To avoid lock bottlenecks this map is dived to several (SHARD_COUNT) map shards.
-type ConcurrentMap []*ConcurrentMapShared
-
-// A "thread" safe string to anything map.
-type ConcurrentMapShared struct {
-	items        map[string]interface{}
+// ShardCount represents the number of shards the map is divided into.
+var ShardCount = 32
+
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
+type ConcurrentMap []*ConcurrentMapShard
+
+// ConcurrentMapShared is a "thread" safe string to anything map.
+type ConcurrentMapShard struct {
+	maxSize uint32
+	idxAdd  uint32
+	idxLast uint32
+	size    uint32
+	mapKeys []string
+	items   map[string]interface{}
 	sync.RWMutex // Read Write mutex, guards access to internal map.
 }
 
-// Creates a new concurrent map.
-func New() ConcurrentMap {
-	m := make(ConcurrentMap, SHARD_COUNT)
-	for i := 0; i < SHARD_COUNT; i++ {
-		m[i] = &ConcurrentMapShared{items: make(map[string]interface{})}
+// New creates a new concurrent map.
+func New(maxSize uint32) ConcurrentMap {
+	m := make(ConcurrentMap, ShardCount)
+	shardSize := maxSize / uint32(ShardCount)
+	if maxSize%uint32(ShardCount) != 0 {
+		shardSize++
+	}
+	for i := 0; i < ShardCount; i++ {
```

Comment:
Should test here if shardSize is 0, otherwise line 76 might break the code

Reply:
done